### DISCUSSION TITLE
Interacting with RPC devices via device files.

### DISCUSSION THREAD

**Discussion by blackdragon2447** [fnuecke/oc2#129] [Q&A] [answered]
I've recently been looking into interacting with RPC devices via devices files, from looking through the devices.py script i was able to decipher that the `/dev/hvc0` file is used to do so, and that this communication is done via jsons separated by null characters, source:
```python
def _write_message(self, data):
    self.file.write(self.MESSAGE_DELIMITER + json.dumps(data) + self.MESSAGE_DELIMITER)
```
so writing `[[null]]{"type": "list"}[[null]]` into the file and than reading the file (using something like cat) should return a json containing a list of all the devices (wrapped in null characters), for how far is this correct (since I cant seem to read anything from the file)?
As for calling methods on devices via this method, do the arguments for the method need to be a json objet:
```json
{
    "type": "invoke",
    "data": {
        "deviceId": 0,
        "name": "getRedstoneInput",
        "parameters": {
            "side": "west"
        }
    }
}
```
or a list:
```json
{
    "type": "invoke",
    "data": {
        "deviceId": 0,
        "name": "getRedstoneInput",
        "parameters": [
            "west"
        ]
    }
}
```
and does the order of the fields on the json matter?

**Comment by sannelo** [comment]
I'm not entirely sure I understand you. 
My friend created a library for c/cpp which just interacts with devices.
https://github.com/Bs0Dd/Coverett

**Reply by blackdragon2447** [reply]
looking at the library it seems to be doing about the same as what i am doing to for some reason my code (rust) refuses to properly write to the device file.

**Comment by fnuecke** [comment, marked as answer]
Yes, that is the basic format of the RPC "protocol". One tricky bit is the need to switch the hvc0 into raw mode *after* opening it (as seen [here](https://github.com/fnuecke/oc2/blob/1.18-forge/src/main/scripts/lib/micropython/devices.py#L61)). The `parameters` field is a list/array, in the order the backing RPC method expects them, so order does matter.
